src/main/java/de/tum/in/www1/artemis/service/hestia/TestwiseCoverageService.java

```diff
@@ -0,0 +1,233 @@
+        this.coverageReportRepository = coverageReportRepository;
+        this.coverageFileReportRepository = coverageFileReportRepository;
+        this.testwiseCoverageReportEntryRepository = testwiseCoverageReportEntryRepository;
+        this.programmingExerciseTestCaseRepository = programmingExerciseTestCaseRepository;
+        this.programmingSubmissionRepository = programmingSubmissionRepository;
+        this.gitService = gitService;
+        this.repositoryService = repositoryService;
+        this.solutionProgrammingExerciseParticipationRepository = solutionProgrammingExerciseParticipationRepository;
+    }
+
+    /**
+     * Creates a coverage report from a testwise coverage report.
+     * Test case names are resolved to a test case of the given programming exercise, adds this reference to the given
+     * entries and saves the entries with the test case reference to the database.
+     * In case, no test case could be found for the given name, the report for this test case will not be saved
+     * @param fileReportByTestCaseName a map containing the test case name as a key and the file coverage reports without the
+     * reference to a test case as a value
+     * @param submission the solution programming submission for which the report is updated
+     * @param exercise the exercise for which the report should be updated
+     */
+    public void createTestwiseCoverageReport(Map<String, Set<CoverageFileReport>> fileReportByTestCaseName, ProgrammingExercise exercise, ProgrammingSubmission submission) {
+        // If the report already exists, do not create a new report. This is the case if the build plan will be re-run
+        boolean reportAlreadyExists = coverageReportRepository.findCoverageReportBySubmissionId(submission.getId()).isPresent();
+        if (reportAlreadyExists) {
+            return;
+        }
+
+        var testCases = programmingExerciseTestCaseRepository.findByExerciseId(exercise.getId());
+        var solutionLineCountByFilePath = getLineCountByFilePath(exercise);
+
+        // Save the full report with the test case and submission, but without the individual file reports as they do not have an ID yet
+        var fullReport = new CoverageReport();
+        fullReport.setSubmission(submission);
+        var savedFullReport = coverageReportRepository.save(fullReport);
+
+        // The file reports unique for a file path. the set is used to aggregate multiple file reports from multiple test
+        // cases into one file report
+        var uniqueFileReports = new HashSet<CoverageFileReport>();
+
+        fileReportByTestCaseName.forEach((testCaseName, fileReports) -> {
+            // retrieve the test matching the extracted test case name
+            var optionalTestCase = testCases.stream().filter(testCase -> testCaseName.equals(testCase.getTestName())).findFirst();
+            if (optionalTestCase.isEmpty()) {
+                return;
```

Comment:
I don't think you should silently ignore this case here. As this method is called directly after the test cases are saved this case would mean that something went wrong somewhere. You should at least log something to the console in these cases, so there is at least some evidence as to what the cause of a wrong CoverageReport is (as this would most likely be noticed by the instructor).